### DISCUSSION TITLE
Brownfield usage of Fabric with Swift

### DISCUSSION THREAD

**Discussion by flokol120** [reactwg/react-native-new-architecture#143] [Q&A] [answered]
I am trying to activate Fabric within an existing iOS app to use Turbo Native Modules. The app was created using Swift, but there are no examples of changes that must be made to use Fabric when using Swift.

What I've tried so far:

1. Follow the documentation to integrate React Native into my existing Swift app using https://reactnative.dev/docs/next/integration-with-existing-apps
2. Translate the changes made to AppDelegate.mm to Swift outlined in https://reactnative.dev/docs/0.70/new-architecture-app-renderer-ios#2-update-your-root-view

After that I ended up with this `AppDelegate.swift`:

```swift
import UIKit
import React
import RCTFabric

@main
class AppDelegate: UIResponder, UIApplicationDelegate, RCTBridgeDelegate {
    
    func sourceURL(for bridge: RCTBridge!) -> URL! {
#if DEBUG
        return URL(string: "http://localhost:8081/index.bundle?platform=ios")
#else
        return Bundle.main.url(forResource: "main", withExtension: "jsbundle")
#endif
    }
    
    static var bridge: RCTBridge?
#if RN_FABRIC_ENABLED
    static var bridgeAdapter: RCTSurfacePresenterBridgeAdapter?
#endif
    var window: UIWindow?
    
#if RN_FABRIC_ENABLED
    var bridgeAdapter: RCTSurfacePresenterBridgeAdapter?
    var reactNativeConfig: Unmanaged<ReactNativeConfig>?
    var contextContainer: Unmanaged<ContextContainer>?
#endif
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.//
#if RN_FABRIC_ENABLED
        contextContainer = Unmanaged.passRetained(ContextContainer())
        reactNativeConfig = Unmanaged.passRetained(EmptyReactNativeConfig())
        
        contextContainer?.takeUnretainedValue().insert("ReactNativeConfig", reactNativeConfig?.takeUnretainedValue())
        
        bridgeAdapter = RCTSurfacePresenterBridgeAdapter(bridge: bridge, contextContainer: contextContainer?.takeUnretainedValue())
        
        AppDelegate.bridge.surfacePresenter = bridgeAdapter?.surfacePresenter
#endif
        return true
    }
    ...
}
```

If I try to build the project after running `USE_FABRIC=1 RCT_NEW_ARCH_ENABLED=1 pod install`, I am getting this error:

```
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "React-RCTFabric-umbrella.h"
        ^
/ios/Pods/Headers/Public/RCTFabric/React-RCTFabric-umbrella.h:13:9: note: in file included from /ios/Pods/Headers/Public/RCTFabric/React-RCTFabric-umbrella.h:13:
#import "React/RCTActivityIndicatorViewComponentView.h"
        ^
/ios/Pods/Headers/Public/React-RCTFabric/React/RCTActivityIndicatorViewComponentView.h:10:9: note: in file included from /ios/Pods/Headers/Public/React-RCTFabric/React/RCTActivityIndicatorViewComponentView.h:10:
#import <React/RCTViewComponentView.h>
        ^
/ios/Pods/Headers/Public/React-RCTFabric/React/RCTViewComponentView.h:10:9: note: in file included from /ios/Pods/Headers/Public/React-RCTFabric/React/RCTViewComponentView.h:10:
#import <React/RCTComponentViewProtocol.h>
        ^
/ios/Pods/Headers/Public/React-RCTFabric/React/RCTComponentViewProtocol.h:10:9: note: in file included from /ios/Pods/Headers/Public/React-RCTFabric/React/RCTComponentViewProtocol.h:10:
#import <react/renderer/componentregistry/ComponentDescriptorProvider.h>
        ^
/ios/Pods/Headers/Public/React-Fabric/react/renderer/componentregistry/ComponentDescriptorProvider.h:10:10: note: in file included from /ios/Pods/Headers/Public/React-Fabric/react/renderer/componentregistry/ComponentDescriptorProvider.h:10:
#include <react/renderer/core/ComponentDescriptor.h>
         ^
/ios/Pods/Headers/Public/React-Fabric/react/renderer/core/ComponentDescriptor.h:10:10: note: in file included from /ios/Pods/Headers/Public/React-Fabric/react/renderer/core/ComponentDescriptor.h:10:
#include <react/renderer/core/EventDispatcher.h>
         ^
/ios/Pods/Headers/Public/React-Fabric/react/renderer/core/EventDispatcher.h:10:10: note: in file included from /ios/Pods/Headers/Public/React-Fabric/react/renderer/core/EventDispatcher.h:10:
#include <react/renderer/core/BatchedEventQueue.h>
         ^
/ios/Pods/Headers/Public/React-Fabric/react/renderer/core/BatchedEventQueue.h:10:10: note: in file included from /ios/Pods/Headers/Public/React-Fabric/react/renderer/core/BatchedEventQueue.h:10:
#include <react/renderer/core/EventQueue.h>
         ^
/ios/Pods/Headers/Public/React-Fabric/react/renderer/core/EventQueue.h:10:10: error: 'memory' file not found
#include <memory>
         ^
/ios/FabricBrownfieldTest/AppDelegate.swift:10:8: error: could not build Objective-C module 'RCTFabric'
import RCTFabric
```

If any more information is needed, I am happy to provide it.

**Comment by cipolleschi** [comment, marked as answer]
Hi @flokol120! Unfortunately, Fabric cannot be used with Swift yet.
There are some architectural changes we need to apply to our C++ layer in order to make it work properly with Swift.

You can create TurboModules in Swift, adding a thin wrapper of ObjectiveC++ though. We have prepared examples [here](https://github.com/react-native-community/RNNewArchitectureLibraries). Look at the table in the README for more information.

**Reply by flokol120** [reply]
Alright, thanks for the quick answer!

**Reply by connorpmullins** [reply]
@cipolleschi - sorry if I'm missing documentation on the topic elsewhere, but what's the state of those architectural changes to the C++ layer? Any open issue I can follow related to the topic?

Thanks!

**Reply by cipolleschi** [reply]
@connorpmullins no, unfortunately there is not an open issue yet.
We are focused on shipping the New Architecture as it is today, for now, so we don't have bandwidth to also work on Swift support (and this makes me really sad).
Hopefully, as soon as the New Arch is out of the door, we will get some bandwidth back to work on those architecture issues.

**Reply by jdbarrera** [reply]
Hi!  Just curious, with the latest RN version release (0.73.6), is it possible to enable fabric in a swift app?  I am able to build successfully but the app crashes right away on the AppDelegate.swift file.

I also noticed there is a stack overflow discussion where they mention making a base class in C++ that is extended in swift, does this apply to the latest RN version? -> https://stackoverflow.com/questions/74038150/can-we-keep-the-appdelegate-swift-file-when-upgrading-from-react-native-0-66-4-t

**Reply by cipolleschi** [reply]
Hi @jdbarrera, thanks for the question.
We are not support ing Swift as of today, because of some structural limitation of the repo we need to work on.

It is not possible to have a base class in C++ and extends that in Swift. Inheritance between languages is not supported by the C++/Swift interop.

For your specific error, I should see the stack trace and the error to be able to help, but it is definitely a very advanced/experimental 
 and unsupported setup.

**Reply by jdbarrera** [reply]
Thanks for the reply @cipolleschi !  Would I be able to convert my swift files to objective-c instead? And then inherit the c++ app delegate and extend it that way?

Also, when Fabric becomes the default, are you planning on having swift support at that time?

**Reply by cipolleschi** [reply]
Hi @jdbarrera!

>  Would I be able to convert my swift files to objective-c instead? And then inherit the c++ app delegate and extend it that way?

Of course, that's possible! That's actually our default setup: the AppDelegate in a new app is an Objective-C++ class which inherit from the `RCTAppDelegate` which is another Objective-C++ class.

Technically, you might be able to even keep your AppDelegate in swift, if you create an indirection layer.

Something like:
`SwiftAppDelegate` --inherits from--> `ObjCAppDelegate` --inherits from--> `RCTAppDelegate`

The important point is that the ObjCAppDelegate must not have any C++/ObjC++ symbol in the header file.

>  when Fabric becomes the default, are you planning on having swift support at that time?

Swift support is something we care about, but we need to do some steps to prepare it. I can't say how long it will take to get there, but it is a topic we discuss regularly, so, as soon as it will be feasible, we will tackle the problem for sure.

**Reply by okwasniewski** [reply]
I think that from 0.74 it might be possible to make Swift inherit directly from RCTAppDelegate.

That's what we are doing in React Native visionOS by default and I've upstreamed some fixes to make it possible in core (I'm not sure if they got picked for the release).

**Reply by jdbarrera** [reply]
Thanks for the info!  I will likely be revisiting this in a week or two so I'll let you know if I have any luck getting it to work

**Comment by liontariai** [comment]
Take a look at https://github.com/microsoft/react-native-test-app , at least the macOS setup uses Swift for the AppDelegate and therefore probably also iOS, I just haven't tested it with iOS. But they do quite some interesting stuff there!

**Comment by vladimirivanoviliev** [comment]
Hey, you can look at the following example I just posted of bridging the Fabric component with Swift:
- https://github.com/reactwg/react-native-new-architecture/discussions/8#discussioncomment-12790617
